### PR TITLE
chore: Go Sql Online Store

### PR DESCRIPTION
**Pull request by kevjumba**
**What this PR does / why we need it**:
Implements a basic sql online store that we can use to perform integration testing within Go instead of requiring entry point to be through Python. (This makes it quite hard to do Go specific testing). Also contains integration tests for the sql online store and the getonlinefeatures grpc service to make sure the correct data is retrieved.

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

**Comment by woop** [comment]
I dont agree with the idea of rewriting online stores into Go. We will end up duplicating a lot of logic that should only live in Python in my opinion. What is the motivation behind this change?

File: sdk/python/tests/integration/online_store/test_universal_online.py [hunk 1]
```diff
@@ -715,7 +715,7 @@ def eventually_apply() -> Tuple[None, bool]:
     assert all(v is None for v in online_features["value"])
 
 
-@pytest.mark.skip
+#@pytest.mark.skip
```
Comment (achals) [inline sdk/python/tests/integration/online_store/test_universal_online.py:718]: prefer to remove the line rather than comment

File: go/internal/feast/sqliteonlinestore.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	if s.db == nil {
+		if s.path == "" {
+			return nil, errors.New("no database path available")
+		}
+		db, err := initializeConnection(s.path)
+		s.db = db
+		if err != nil {
+			return nil, err
+		}
+	}
```
Comment (achals) [inline go/internal/feast/sqliteonlinestore.go:143]: This should probably be guarded by a lock to prevent concurrent access creating multiple dangling connections

File: go/internal/feast/sqliteonlinestore.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+		defer rows.Close()
+		for rows.Next() {
+			var entity_key string
+			var feature_name string
+			var valueString string
+			var event_ts time.Time
+			var value types.Value
+			err = rows.Scan(&entity_key, &feature_name, &valueString, &event_ts)
+			if err != nil {
+				log.Fatal(err)
```
Comment (achals) [inline go/internal/feast/sqliteonlinestore.go:100]: Why are something logged and other things returned as `fmt.Errorf`? Also a `log.Fatal` will crash the running process which is something we should definietly not do in this code path - that should be reserved for unrecoverable errors like invalid config at start up time or something.

File: go/internal/feast/sqliteonlinestore.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+			results[entityNameToEntityIndex[entity_key]][featureNamesToIdx[feature_name]] = FeatureData{reference: serving.FeatureReferenceV2{FeatureViewName: featureViewName, FeatureName: feature_name},
+				timestamp: *timestamppb.New(event_ts),
+				value:     types.Value{Val: value.Val},
+			}
+		}
+	}
+	return results, nil
+}
+
+func (s *SqliteOnlineStore) Update(ctx context.Context, config *RepoConfig, tables_to_delete []*FeatureView, tables_to_keep []*FeatureView) error {
```
Comment (achals) [inline go/internal/feast/sqliteonlinestore.go:114]: Do you need this if we never write rows to the sqlite database?

**Comment by kevjumba** [comment]
> I dont agree with the idea of rewriting online stores into Go. We will end up duplicating a lot of logic that should only live in Python in my opinion. What is the motivation behind this change?

Adding @pyalex and @achals. Oleksii raised the concern of difficulties in writing certain tests for the Go feature server specifically and we talked about it and decided to use sqlite as a path as a simple testing backbone to do certain integration/unit testing.

File: go/internal/feast/sqliteonlinestore.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	path    string
+	db      *sql.DB
+}
+
+// Creates a new sqlite online store object. onlineStoreConfig should have relative path of database file with respect to repoConfig.repoPath.
+func NewSqliteOnlineStore(project string, repoConfig *RepoConfig, onlineStoreConfig map[string]interface{}) (*SqliteOnlineStore, error) {
+	store := SqliteOnlineStore{project: project}
+	if db_path, ok := onlineStoreConfig["path"]; !ok {
+		return nil, fmt.Errorf("cannot find sqlite path %s", db_path)
+	} else if dbPathStr, ok := db_path.(string); !ok {
```
[the remark this answers is not in the document]
Comment (kevjumba) [reply go/internal/feast/sqliteonlinestore.go:32]: Changed.

File: sdk/python/tests/integration/online_store/test_universal_online.py (hunk 1 above)
Comment (kevjumba) [reply sdk/python/tests/integration/online_store/test_universal_online.py:718]: Oops this was an accident.

File: go/internal/feast/sqliteonlinestore.go (hunk 3 above)
Comment (kevjumba) [reply go/internal/feast/sqliteonlinestore.go:100]: This was part of my testing code and accidentally got left in there, thanks for the catch.

File: go/internal/feast/sqliteonlinestore.go (hunk 4 above)
Comment (kevjumba) [reply go/internal/feast/sqliteonlinestore.go:114]: I'll remove it for now.

File: go/internal/feast/sqliteonlinestore.go (hunk 2 above)
Comment (kevjumba) [reply go/internal/feast/sqliteonlinestore.go:143]: added a mutex.

**Comment by woop** [comment]
> > I dont agree with the idea of rewriting online stores into Go. We will end up duplicating a lot of logic that should only live in Python in my opinion. What is the motivation behind this change?
> 
> Adding @pyalex and @achals. Oleksii raised the concern of difficulties in writing certain tests for the Go feature server specifically and we talked about it and decided to use sqlite as a path as a simple testing backbone to do certain integration/unit testing.

Can you be more specific on the types of testing that isn't possible? Is the idea that we would reimplement all these `update` methods from the online store interface and make the Go implementation more than a reader/writer?

**Comment by achals** [comment]
> > > I dont agree with the idea of rewriting online stores into Go. We will end up duplicating a lot of logic that should only live in Python in my opinion. What is the motivation behind this change?
> 
> > 
> 
> > Adding @pyalex and @achals. Oleksii raised the concern of difficulties in writing certain tests for the Go feature server specifically and we talked about it and decided to use sqlite as a path as a simple testing backbone to do certain integration/unit testing.
> 
> 
> 
> Can you be more specific on the types of testing that isn't possible? Is the idea that we would reimplement all these `update`/`destroy` methods from the online store interface and make the Go implementation more than a reader/writer?

I believe the idea was to have only the read path implemented for SQLite to make it easy to run a full test in pure go using a prebuilt SQLite file without crossing the boundary from python to go, or spinning up redis and writing to it before testing the go code that read it. 

I agree that the update method should not be included in this pr, unless I'm missing something.

File: go.mod [hunk 6]
```diff
@@ -24,6 +24,9 @@ require (
 	github.com/klauspost/compress v1.13.6 // indirect
 	github.com/kr/pretty v0.1.0 // indirect
 	github.com/pierrec/lz4/v4 v4.1.9 // indirect
+	github.com/go-sql-driver/mysql v1.6.0 // indirect
```
Comment (pyalex) [inline go.mod:27]: why do we need mysql?

File: go/internal/feast/sqliteonlinestore.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	entityNameToEntityIndex := make(map[string]int)
+	in_query := make([]string, len(entityKeys))
+	serialized_entities := make([]interface{}, len(entityKeys))
+	for i := 0; i < len(entityKeys); i++ {
+		serKey, err := serializeEntityKey(&entityKeys[i])
+		if err != nil {
+			return nil, err
+		}
+		// TODO: fix this, string conversion is not safe
+		entityNameToEntityIndex[string(*serKey)] = i
```
Comment (pyalex) [inline go/internal/feast/sqliteonlinestore.go:73]: Please replace `string(..)` with `hash(..)`

File: go/internal/feast/sqliteonlinestore.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+									ORDER BY entity_key`, tableId(project, featureViewName), strings.Join(in_query, ","))
+		rows, err := s.db.Query(query_string, serialized_entities...)
+		if err != nil {
+			return nil, err
+		}
+		defer rows.Close()
+		for rows.Next() {
+			var entity_key string
+			var feature_name string
+			var valueString string
```
Comment (pyalex) [inline go/internal/feast/sqliteonlinestore.go:99]: can we read bytes from sqlite? seems like we can loose data here as well

File: sdk/python/tests/integration/online_store/test_universal_online.py [hunk 9]
```diff
@@ -714,6 +714,7 @@ def eventually_apply() -> Tuple[None, bool]:
     assert all(v is None for v in online_features["value"])
 
 
+@pytest.mark.skip
```
Comment (pyalex) [inline sdk/python/tests/integration/online_store/test_universal_online.py:717]: some issue with rebase probably? shouldn't be skipped anymore

File: sdk/python/feast/templates/local/example.py [hunk 10]
```diff
@@ -8,7 +8,7 @@
 # production, you can use your favorite DWH, such as BigQuery. See Feast documentation
 # for more info.
 driver_hourly_stats = FileSource(
-    path="%PARQUET_PATH%",
+    path="data/driver_stats.parquet",
```
Comment (pyalex) [inline sdk/python/feast/templates/local/example.py:11]: Isn't this an template?

File: go/internal/feast/sqliteonlinestore.go [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+									FROM %s
+									WHERE entity_key IN (%s)
+									ORDER BY entity_key`, tableId(project, featureViewName), strings.Join(in_query, ","))
+		rows, err := s.db.Query(query_string, serialized_entities...)
+		if err != nil {
+			return nil, err
+		}
+		defer rows.Close()
+		for rows.Next() {
+			var entity_key string
```
Comment (pyalex) [inline go/internal/feast/sqliteonlinestore.go:97]: same comment as below

File: go/internal/feast/sqliteonlinestore.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+	}
+	featureNamesToIdx := make(map[string]int)
+	for idx, name := range featureNames {
+		featureNamesToIdx[name] = idx
+	}
+
+	for idx := 0; idx < len(entityKeys); idx++ {
+		results[idx] = make([]FeatureData, featureCount)
+	}
+	for _, featureViewName := range featureViewNames {
```
Comment (pyalex) [inline go/internal/feast/sqliteonlinestore.go:86]: do we need this for loop here? Can't we just query all keys in one go?

Comment (pyalex) [reply go/internal/feast/sqliteonlinestore.go:86]: Oh, I see you have table per feature view. I guess that's fine for sqlite implementation.

File: sdk/python/tests/integration/online_store/test_universal_online.py (hunk 9 above)
Comment (kevjumba) [reply sdk/python/tests/integration/online_store/test_universal_online.py:717]: Fixed.

File: go.mod (hunk 6 above)
Comment (kevjumba) [reply go.mod]: Fixed.

File: go/internal/feast/sqliteonlinestore.go (hunk 7 above)
Comment (kevjumba) [reply go/internal/feast/sqliteonlinestore.go]: Added a function that hashes byte [] to hexcode using sha1.

File: go/internal/feast/sqliteonlinestore.go (hunk 12 above)
Comment (kevjumba) [reply go/internal/feast/sqliteonlinestore.go:87]: Yes, I am adhering to the API set by the online store API.

File: go/internal/feast/sqliteonlinestore.go (hunk 11 above)
Comment (kevjumba) [reply go/internal/feast/sqliteonlinestore.go]: Fixed above.

File: go/internal/feast/sqliteonlinestore.go (hunk 8 above)
Comment (kevjumba) [reply go/internal/feast/sqliteonlinestore.go]: Changed to byte []

File: sdk/python/feast/templates/local/example.py (hunk 10 above)
Comment (kevjumba) [reply sdk/python/feast/templates/local/example.py]: Fixed.

File: go/internal/test/feature_repo/example.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+from datetime import timedelta
+
+from feast import Entity, Feature, FeatureView, FileSource, ValueType
+
+# Read data from parquet files. Parquet is convenient for local development mode. For
+# production, you can use your favorite DWH, such as BigQuery. See Feast documentation
+# for more info.
+driver_hourly_stats = FileSource(
+    path="/Users/kevinzhang/tecton-ai/offline_store/feast/go/internal/test/feature_repo/data/driver_stats.parquet",
```
[the remark this answers is not in the document]
Comment (kevjumba) [reply go/internal/test/feature_repo/example.py]: Fixed.

**Comment by kevjumba** [comment]
@everyone, new commits right now are for debugging ci go tests so please ignore.

File: go/internal/feast/featurestore.go [hunk 14]
```diff
@@ -104,6 +105,8 @@ func (fs *FeatureStore) GetOnlineFeatures(
 
 	numRows, err := fs.validateEntityValues(entityProtos)
 	if err != nil {
+		log.Println("1")
```
Comment (pyalex) [inline go/internal/feast/featurestore.go]: what's that for?

File: go/internal/feast/sqliteonlinestore_test.go [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+	assert.Nil(t, err)
+	returnedFeatureValues := make([]*types.Value, 0)
+	returnedFeatureNames := make([]string, 0)
+	for _, featureVector := range featureData {
+		for idx := range featureVector {
+			returnedFeatureValues = append(returnedFeatureValues, &featureVector[idx].value)
+			returnedFeatureNames = append(returnedFeatureNames, featureVector[idx].reference.FeatureName)
+		}
+	}
+	expectedFeatureValues := []*types.Value{
```
Comment (pyalex) [inline go/internal/feast/sqliteonlinestore_test.go:60]: just a suggestion, can we maybe read this values from parquet file? It's not really clear why these values are expected

Comment (pyalex) [reply go/internal/feast/sqliteonlinestore_test.go:60]: I assumed that this sqlite db was materialized from parquet file, so we can read from the source.
If it's not true, than it's fine

File: go/internal/feast/sqliteonlinestore.go [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+	db, err := sql.Open("sqlite3", db_path)
+	if err != nil {
+		return nil, err
+	}
+	return db, nil
+}
+
+func hashSerializedEntityKey(serializedEntityKey *[]byte) string {
+	h := sha1.New()
+	h.Write(*serializedEntityKey)
```
Comment (pyalex) [inline go/internal/feast/sqliteonlinestore.go:152]: please check the pointer before dereferencing it

File: go/cmd/server/server_test.go [hunk 17]
```diff
@@ -58,21 +66,51 @@ func getClient(ctx context.Context) (serving.ServingServiceClient, func()) {
...
+		{Val: &types.Value_Int64Val{Int64Val: 1001}},
+		{Val: &types.Value_Int64Val{Int64Val: 1003}},
+		{Val: &types.Value_Int64Val{Int64Val: 1005}},
+	}
+	expectedFeatureNamesResp := []string{"driver_id", "conv_rate", "acc_rate", "avg_daily_trips"}
 	assert.Nil(t, err)
 	assert.NotNil(t, response)
+
+	// Columnar so first column should have column names of all features
+	assert.True(t, reflect.DeepEqual(response.Results[0].Values, expectedEntityValuesResp))
```
Comment (pyalex) [inline go/cmd/server/server_test.go:110]: let's add here check for feature values as well, since we read them from the source file

File: go/internal/feast/sqliteonlinestore.go (hunk 16 above)
Comment (kevjumba) [reply go/internal/feast/sqliteonlinestore.go:152]: Done.

File: go/cmd/server/server_test.go (hunk 17 above)
Comment (kevjumba) [reply go/cmd/server/server_test.go:127]: DOne.

File: go/internal/feast/sqliteonlinestore_test.go (hunk 15 above)
Comment (kevjumba) [reply go/internal/feast/sqliteonlinestore_test.go]: Don.e

File: .github/workflows/unit_tests.yml [hunk 18]
```diff
@@ -76,6 +76,7 @@ jobs:
       - name: Upgrade pip version
         run: |
           pip install --upgrade "pip>=21.3.1"
+          pip install pip-tools
```
Comment (achals) [inline .github/workflows/unit_tests.yml]: Why is this needed here? Can you pin the version?

File: Makefile [hunk 19]
```diff
@@ -142,7 +142,10 @@ compile-go-lib: install-go-proto-dependencies install-go-ci-dependencies
 	cd sdk/python && python setup.py build_go_lib
 
 test-go: compile-protos-go
+	pip install feast
+	cd go/internal/test && feast init feature_repo && cd feature_repo && feast apply && feast materialize-incremental $(shell date -u +"%Y-%m-%dT%H:%M:%S") && cd ../../../
 	go test ./...
```
Comment (achals) [inline Makefile:147]: I'd recommend splitting this into more lines. Also, each line is executed in isolation, so the `cd ../../../` isn't needed.:
```
    cd go/internal/test && feast init feature_repo 
    cd go/internal/test/feature_repo && feast apply && feast materialize-incremental $(shell date -u +"%Y-%m-%dT%H:%M:%S")
    go test ./...
```

File: go/cmd/server/server_test.go [hunk 20]
```diff
@@ -58,21 +67,67 @@ func getClient(ctx context.Context) (serving.ServingServiceClient, func()) {
 }
 
 func TestGetFeastServingInfo(t *testing.T) {
-	t.Skip("@todo(achals): feature_repo isn't checked in yet")
 	ctx := context.Background()
-	client, closer := getClient(ctx)
+	// Pregenerated using `feast init`.
+	dir := "../../internal/test"
```
Comment (achals) [inline go/cmd/server/server_test.go:72]: I think this is pretty fragile.. having the feature repo created in a separate command from the `go test` command. Can we have a `Setup` method create the feature repo here so that everything is self-contained and it gets executed as part of this test?

File: go/internal/feast/featurestore.go [hunk 21]
```diff
@@ -130,6 +130,7 @@ func (fs *FeatureStore) GetOnlineFeatures(
 
 	entityNameToJoinKeyMap, expectedJoinKeysSet, err := fs.getEntityMaps(requestedFeatureViews)
 	if err != nil {
+
```
Comment (achals) [inline go/internal/feast/featurestore.go:133]: rm newline

File: go/internal/feast/onlinestore.go [hunk 22]
```diff
@@ -33,7 +34,7 @@ type OnlineStore interface {
 	// Feature object as pointers in GetOnlineFeaturesResponse)
 	// => allocate memory for each field once in OnlineRead
 	// and reuse them in GetOnlineFeaturesResponse?
-	OnlineRead(ctx context.Context, entityKeys []types.EntityKey, featureViewNames []string, featureNames []string) ([][]FeatureData, error)
+	OnlineRead(ctx context.Context, entityKeys []*types.EntityKey, featureViewNames []string, featureNames []string) ([][]FeatureData, error)
```
Comment (achals) [inline go/internal/feast/onlinestore.go:37]: Why are we changing the interface here?

File: go/internal/feast/onlinestore.go [hunk 23]
```diff
@@ -51,12 +52,13 @@ func getOnlineStoreType(onlineStoreConfig map[string]interface{}) (string, bool)
 func NewOnlineStore(config *RepoConfig) (OnlineStore, error) {
 	onlineStoreType, ok := getOnlineStoreType(config.OnlineStore)
 	if !ok {
-		return nil, fmt.Errorf("could not get online store type from online store config: %+v", config.OnlineStore)
+		onlineStore, err := NewSqliteOnlineStore(config.Project, config, config.OnlineStore)
+		return onlineStore, err
```
Comment (achals) [inline go/internal/feast/onlinestore.go:56]: This seems wrong ? DOn't we need `ok && onlineStoreType == "sqlite"` before we do this code path?

File: go/internal/feast/onlinestore.go (hunk 22 above)
Comment (kevjumba) [reply go/internal/feast/onlinestore.go:37]: I think it doesn't make sense to have to copy the EntityKey around, changing it to a pointer doesn't change the functionality, just makes it more efficient

File: go/internal/feast/onlinestore.go (hunk 23 above)
Comment (kevjumba) [reply go/internal/feast/onlinestore.go:56]: In the python implementation, no type for online store  config automatically assumes sqlite as the onlinestore => I believe I tried adding type = sqlite there were some small config collisions so I decided it makes sense to adhere to the python online store configuraiton.

File: go/internal/feast/featurestore.go (hunk 21 above)
Comment (kevjumba) [reply go/internal/feast/featurestore.go:133]: Fixed.

File: .github/workflows/unit_tests.yml (hunk 18 above)
Comment (kevjumba) [reply .github/workflows/unit_tests.yml]: Removed.

File: go/cmd/server/server_test.go (hunk 20 above)
Comment (kevjumba) [reply go/cmd/server/server_test.go]: Done, added methods for setup and cleanup

File: Makefile (hunk 19 above)
Comment (kevjumba) [reply Makefile]: Removed all of it

File: go/internal/test/utils.go [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+
+	if err != nil {
+		return err
+	}
+	cmd.Dir = path
+	err = cmd.Run()
+	if err != nil {
+		log.Println("in3")
+		return err
+	}
```
Comment (achals) [inline go/internal/test/utils.go:116]: rm?

File: go/internal/test/utils.go [hunk 25]
```diff
@@ -0,0 +1,154 @@
...
+	feature_repo_path, err := filepath.Abs(filepath.Join(path, "feature_repo"))
+	if err != nil {
+		return err
+	}
+	applyCommand.Dir = feature_repo_path
+	err = applyCommand.Run()
+	if err != nil {
+		log.Println("in1")
+		return err
+	}
```
Comment (achals) [inline go/internal/test/utils.go:128]: rm?

File: go/internal/test/utils.go [hunk 26]
```diff
@@ -0,0 +1,154 @@
...
+		t.Year(), t.Month(), t.Day(),
+		t.Hour(), t.Minute(), t.Second())
+	materializeCommand := exec.Command("feast", "materialize-incremental", formattedTime)
+	materializeCommand.Env = os.Environ()
+	materializeCommand.Dir = feature_repo_path
+	err = materializeCommand.Run()
+	if err != nil {
+		log.Println("in2")
+		return err
+	}
```
Comment (achals) [inline go/internal/test/utils.go:140]: rm?

File: go/internal/test/utils.go [hunk 27]
```diff
@@ -0,0 +1,154 @@
...
+				}
+			} else {
+				correctFeatureRows[Row.Driver_id] = Row
+			}
+		}
+	}
+	return correctFeatureRows
+}
+
+func SetupFeatureRepo(basePath string) error {
```
Comment (achals) [inline go/internal/test/utils.go:103]: If you wanna, you can also use TestSuites to attach this method to the test struct:

https://pkg.go.dev/github.com/stretchr/testify/suite

File: go/internal/feast/utils.go [hunk 28]
```diff
@@ -0,0 +1,96 @@
+package feast
```
Comment (achals) [inline go/internal/feast/utils.go:1]: nit, `utils.go` is smell that this is a kitchen sink - can we name this file something more descriptive?

File: go/internal/test/utils.go [hunk 29]
```diff
@@ -0,0 +1,154 @@
...
+type Row struct {
+	Event_timestamp int64   `json:"Event_timestamp"`
+	Driver_id       int64   `json:"Driver_id"`
+	Conv_rate       float32 `json:"Conv_rate"`
+	Acc_rate        float32 `json:"Acc_rate"`
+	Avg_daily_trips int32   `json:"Avg_daily_trips"`
+	Created         int64   `json:"Created"`
+}
+
+func ReadParquet(filePath string) ([]*Row, error) {
```
Comment (achals) [inline go/internal/test/utils.go:52]: Is this method duplicated?

File: go/internal/test/utils.go [hunk 30]
```diff
@@ -0,0 +1,154 @@
...
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/xitongsys/parquet-go-source/local"
+	"github.com/xitongsys/parquet-go/reader"
+	"github.com/xitongsys/parquet-go/source"
+)
+
+var jsonSchema string = `
```
Comment (achals) [inline go/internal/test/utils.go:17]: Any reason why so much stuff is duplicated here?

File: go/internal/test/utils.go [hunk 31]
```diff
@@ -0,0 +1,154 @@
+package test
```
Comment (achals) [inline go/internal/test/utils.go:1]: same comment about the file name.

**Review by achals — CHANGES_REQUESTED** [review]

File: go/internal/test/utils.go (hunk 29 above)
Comment (kevjumba) [reply go/internal/test/utils.go]: Should have deleted utils.go in feast, not sure why it's not...

File: go/internal/test/utils.go (hunk 24 above)
Comment (kevjumba) [reply go/internal/test/utils.go]: Shouldve been included in clean up commit, not sure why it wasnt, removed.

File: go/internal/test/utils.go (hunk 26 above)
Comment (kevjumba) [reply go/internal/test/utils.go]: Same

File: go/internal/test/utils.go (hunk 31 above)
Comment (kevjumba) [reply go/internal/test/utils.go]: Fixed.

File: go/internal/test/go_integration_test_utils.go [hunk 32]
```diff
@@ -0,0 +1,151 @@
...
+	}
+	applyCommand.Dir = feature_repo_path
+	err = applyCommand.Run()
+	if err != nil {
+		return err
+	}
+	t := time.Now()
+	formattedTime := fmt.Sprintf("%d-%02d-%02dT%02d:%02d:%02d",
+		t.Year(), t.Month(), t.Day(),
+		t.Hour(), t.Minute(), t.Second())
```
Comment (achals) [inline go/internal/test/go_integration_test_utils.go:130]: You should be able to just do
```
current_time.Format(time.RFC3339Nano)
```

File: Makefile [hunk 33]
```diff
@@ -141,7 +141,9 @@ compile-go-lib: install-go-proto-dependencies install-go-ci-dependencies
 	python -m pip install pybindgen==0.22.0
 	cd sdk/python && python setup.py build_go_lib
 
+# Needs feast package to setup the feature store
 test-go: compile-protos-go
+	pip install feast
```
Comment (pyalex) [inline Makefile:146]: Can we install it from the repo, not from pypi?

File: go/internal/test/utils.go (hunk 25 above)
Comment (kevjumba) [reply go/internal/test/utils.go]: Fixed.

File: sdk/python/setup.py [hunk 34]
```diff
@@ -101,7 +101,7 @@
 
 CI_REQUIRED = (
     [
-        "cryptography==3.4.8",
+        "cryptography==3.3.2",
```
Comment (achals) [inline sdk/python/setup.py:104]: why are we downgrading this?

**Review by achals — APPROVED** [review]
/lgtm
